rgw: make RGWEnv return a const ref. to its map

theanalyst opened the pull request:
We already have a public method `set` for setting values in the RGW env
map, making the map as a read only. In addition:

- Added const to other  methods in `RGWEnv` which are getters
- `req_info::init_meta` also reused the same `iter` name for an internal
  find, changed the var name
- `add_grants_headers` now accepts an RGWEnv instead of the map
- use range based for loops wherever the code was changed\
- req_info holds a ptr to a const RGWEnv, since we seem to only read the
  values from it

Signed-off-by: Abhishek Lekshmanan <abhishek@suse.com>

rzarzynski commented on src/rgw/rgw_common.cc:
```diff
@@ -201,7 +201,7 @@ static string get_abs_path(const string& request_uri) {
 }
 
-req_info::req_info(CephContext *cct, class RGWEnv *e) : env(e) {
+req_info::req_info(CephContext *cct, class RGWEnv *_env) : env(_env) {
```
The underscore in `_env` [may be omitted](https://stackoverflow.com/a/6185043):
```cpp
req_info::req_info(CephContext *cct, class RGWEnv *env) : env(env) {
```
A bit more const-correct variant would look like:
```cpp
req_info::req_info(CephContext* const cct, class RGWEnv* const env) : env(env) {
```

rzarzynski replied:
Just a suggestion: maybe `req_info::env` could be turned into a pointer to constant. If I see correctly the methods of `req_info` call only `RGWEnv::get`.

theanalyst replied:
Yeah I thought of going that route, and got around to fixing all the places where we mutate the map without the methods. I also vote for this.

rzarzynski commented on src/rgw/rgw_common.cc:
```diff
@@ -384,12 +384,10 @@ void req_info::init_meta_info(bool *found_bad_meta)
-  map<string, string, ltstr_nocase>& m = env->get_map();
-  map<string, string, ltstr_nocase>::iterator iter;
-  for (iter = m.begin(); iter != m.end(); ++iter) {
+  for (const auto& iter: env->get_map()) {
```
Hmm, technically it isn't an iterator anymore. Maybe just `kv` instead?

rzarzynski commented on src/rgw/rgw_common.cc:
```diff
@@ -427,8 +424,8 @@ void req_info::init_meta_info(bool *found_bad_meta)
   }
-  for (iter = x_meta_map.begin(); iter != x_meta_map.end(); ++iter) {
-    dout(10) << "x>> " << iter->first << ":" << rgw::crypt_sanitize::x_meta_map{iter->first, iter->second} << dendl;
+  for (const auto& iter: x_meta_map) {
```
The same as above (I stop mentioning that).

theanalyst replied:
ack

rzarzynski commented on src/rgw/rgw_common.h:
```diff
@@ -383,16 +383,16 @@ class RGWEnv {
   void set(const boost::string_ref& name, const boost::string_ref& val);
-  const char *get(const char *name, const char *def_val = NULL);
-  int get_int(const char *name, int def_val = 0);
+  const char *get(const char *name, const char *def_val = NULL) const;
```
As we're already touching the line it might be worth to switch to `nullptr` as well.

theanalyst replied:
hm sure

rzarzynski commented on src/rgw/rgw_common.cc:
```diff
@@ -412,10 +410,9 @@ void req_info::init_meta_info(bool *found_bad_meta)
+        auto it = x_meta_map.find(name_low);
+        if (it != x_meta_map.end()) {
+          string old = it->second;
           int pos = old.find_last_not_of(" \t"); /* get rid of any whitespaces after the value */
```
May we have `size_t` in place of `int`?

rzarzynski commented on src/rgw/rgw_rest_client.cc:
```diff
@@ -264,15 +261,14 @@ int RGWRESTSimpleRequest::forward_request(RGWAccessKey& key, req_info& info, siz
-    headers.push_back(pair<string, string>(iter->first, iter->second));
+  const auto & m = new_env.get_map();
+  for (const auto& iter: m) {
+    headers.push_back(pair<string, string>(iter.first, iter.second));
```
`std::make_pair` could be helpful here.

cbodley replied:
`headers.emplace_back(iter)` works here as well

theanalyst replied:
ack

rzarzynski commented on src/rgw/rgw_common.cc:
```diff
@@ -412,10 +410,9 @@ void req_info::init_meta_info(bool *found_bad_meta)
+        if (it != x_meta_map.end()) {
+          string old = it->second;
           int pos = old.find_last_not_of(" \t"); /* get rid of any whitespaces after the value */
           old = old.substr(0, pos + 1);
```
I know it's not directly related to the proposed changes but we might have an issue here. AFAIK `std::string::npos` *usually* is the max of `size_t`, so adding `1` to it would give us `0`. If so, the `substr(0, 0)` eradicates the old value. This isn't coherent with the comment above if my understanding is correct.

theanalyst replied:
checking for pos != string::npos (or use boost::string::trim_right..)

rzarzynski replied:
I vote for `trim_right`. We're already using `boost::algorithm::trim` in couple of places.

theanalyst commented on src/rgw/rgw_common.h:
```diff
@@ -383,16 +383,16 @@ class RGWEnv {
-  int get_int(const char *name, int def_val = 0);
+  const char *get(const char *name, const char *def_val = NULL) const;
+  int get_int(const char *name, int def_val = 0) const;
   bool get_bool(const char *name, bool def_val = 0);
```
this should also be const if we go for const ptr

theanalyst commented on src/rgw/rgw_rest_client.cc:
```diff
@@ -486,16 +482,15 @@ int RGWRESTStreamWriteRequest::put_obj_init(RGWAccessKey& key, rgw_obj& obj, uin
-  for (iter = m.begin(); iter != m.end(); ++iter) {
-    headers.push_back(pair<string, string>(iter->first, iter->second));
+  for (const auto& iter: new_env.get_map()) {
+    headers.push_back(pair<string, string>(iter.first, iter.second));
```
emplace

theanalyst replied:
done

theanalyst commented on src/rgw/rgw_rest_conn.cc:
```diff
@@ -180,17 +180,17 @@ int RGWRESTConn::get_obj(const rgw_user& uid, req_info *info /* optional */, rgw
 #define SEARCH_AMZ_PREFIX "HTTP_X_AMZ_"
-    for (map<string, string>::iterator iter = orig_map.lower_bound(SEARCH_AMZ_PREFIX); iter != orig_map.end(); ++iter) {
-      const string& name = iter->first;
+    for (const auto& iter: orig_map) {
```
hm this doesn't  look right, use the lower bound, and maybe #define -> constexpr?

theanalyst commented on src/rgw/rgw_env.cc:
```diff
@@ -62,36 +62,35 @@ int RGWEnv::get_int(const char *name, int def_val)
 
   const char *s = iter->second.c_str();
-  return atoll(s);  
+  return atoll(s);
```
stol?

theanalyst replied:
since stol throws have to catch the exceptions as well

theanalyst commented:
will squash the PRs after reviews

yehudasa commented on src/rgw/rgw_rest_conn.cc:
```diff
@@ -183,14 +183,14 @@ int RGWRESTConn::get_obj(const rgw_user& uid, req_info *info /* optional */, rgw
       if (name == "HTTP_X_AMZ_DATE") /* dont forward date from original request */
         continue;
-      if (name.compare(0, sizeof(SEARCH_AMZ_PREFIX) - 1, "HTTP_X_AMZ_") != 0)
+      if (name.compare(0, sizeof(SEARCH_AMZ_PREFIX) - 1, SEARCH_AMZ_PREFIX) != 0)
```
let's make sure that compiler actually optimizes it correctly before changing it

rzarzynski replied:
The test program:
```cpp
#include <cstring>
#include <iostream>

int main () {
  static constexpr char test_string[] = "Just a test C\0string";
  std::cout << "the length is: " << std::strlen(test_string) << std::endl;
}
```
Verification with GCC 4.8.4 on AMD64:
```
$ g++ -std=c++11 /tmp/test.cc -o /tmp/test
$ /tmp/test 
the length is: 13
$ objdump -dC /tmp/test

...

000000000040089d <main>:
  40089d:	55                   	push   %rbp
  40089e:	48 89 e5             	mov    %rsp,%rbp
  4008a1:	be c1 09 40 00       	mov    $0x4009c1,%esi
  4008a6:	bf 80 10 60 00       	mov    $0x601080,%edi
  4008ab:	e8 c0 fe ff ff       	callq  400770 <std::basic_ostream<char, std::char_traits<char> >& std::operator<< <std::char_traits<char> >(std::basic_ostream<char, std::char_traits<char> >&, char const*)@plt>
  4008b0:	be 0d 00 00 00       	mov    $0xd,%esi
  4008b5:	48 89 c7             	mov    %rax,%rdi
  4008b8:	e8 c3 fe ff ff       	callq  400780 <std::ostream::operator<<(unsigned long)@plt>
  4008bd:	be a0 07 40 00       	mov    $0x4007a0,%esi
  4008c2:	48 89 c7             	mov    %rax,%rdi
  4008c5:	e8 c6 fe ff ff       	callq  400790 <std::ostream::operator<<(std::ostream& (*)(std::ostream&))@plt>
  4008ca:	b8 00 00 00 00       	mov    $0x0,%eax
  4008cf:	5d                   	pop    %rbp
  4008d0:	c3                   	retq 

...

```

yehudasa replied:
+1

theanalyst commented on src/rgw/rgw_rest_conn.cc:
```diff
@@ -183,14 +183,14 @@ int RGWRESTConn::get_obj(const rgw_user& uid, req_info *info /* optional */, rgw
-#define SEARCH_AMZ_PREFIX "HTTP_X_AMZ_"
-    for (const auto& iter: orig_map) {
-      const string& name = iter.first;
+    #define SEARCH_AMZ_PREFIX "HTTP_X_AMZ_"
```
ack

theanalyst commented:
test this please

theanalyst commented:
@rzarzynski updated

rzarzynski commented on src/rgw/rgw_rest_client.cc:
```diff
@@ -454,7 +450,7 @@ int RGWRESTStreamWriteRequest::put_obj_init(RGWAccessKey& key, rgw_obj& obj, uin
   new_info.request_uri = new_info.script_uri;
 
-  map<string, string, ltstr_nocase>& m = new_env.get_map();
+
```
Do we need this empty line?

theanalyst replied:
actually the next line also can be squashed if go for a range based for loop, pushing the changeset. I can always push back the previous version if we feel this is not necessary

rzarzynski commented on src/rgw/rgw_rest_client.cc:
```diff
@@ -264,15 +261,14 @@ int RGWRESTSimpleRequest::forward_request(RGWAccessKey& key, req_info& info, siz
-  for (iter = m.begin(); iter != m.end(); ++iter) {
-    headers.push_back(pair<string, string>(iter->first, iter->second));
+  const auto & m = new_env.get_map();
+  for (const auto& iter: m) {
```
`iter` might be a bit confusing here. Maybe:
```cpp
  for (const auto& kv: new_env.get_map()) {
    headers.emplace_back(kv);
  }

theanalyst replied:
thanks, more compact, did this in one other place, forgot here

rzarzynski commented on src/rgw/rgw_rest_client.cc:
```diff
@@ -264,15 +261,14 @@ int RGWRESTSimpleRequest::forward_request(RGWAccessKey& key, req_info& info, siz
   map<string, string>& meta_map = new_info.x_meta_map;
-  for (iter = meta_map.begin(); iter != meta_map.end(); ++iter) {
-    headers.push_back(pair<string, string>(iter->first, iter->second));
+  for (const auto& iter: meta_map) {
```
`iter` -> `kv`.

rzarzynski commented on src/rgw/rgw_rest_client.cc:
```diff
@@ -264,15 +261,14 @@ int RGWRESTSimpleRequest::forward_request(RGWAccessKey& key, req_info& info, siz
-  for (iter = meta_map.begin(); iter != meta_map.end(); ++iter) {
-    headers.push_back(pair<string, string>(iter->first, iter->second));
+  for (const auto& iter: meta_map) {
+    headers.push_back(pair<string, string>(iter.first, iter.second));
```
`emplace_back(kv)`

theanalyst replied:
good catch

rzarzynski commented on src/rgw/rgw_rest_client.cc:
```diff
@@ -486,16 +482,15 @@ int RGWRESTStreamWriteRequest::put_obj_init(RGWAccessKey& key, rgw_obj& obj, uin
-  map<string, string>::iterator iter;
-  for (iter = m.begin(); iter != m.end(); ++iter) {
-    headers.push_back(pair<string, string>(iter->first, iter->second));
+  for (const auto& iter: new_env.get_map()) {
```
`iter` -> `kv`.

rzarzynski commented on src/rgw/rgw_rest_client.cc:
```diff
@@ -680,10 +675,8 @@ int RGWRESTStreamRWRequest::send_request(RGWAccessKey *key, map<string, string>&
-  map<string, string>::iterator iter;
-  for (iter = m.begin(); iter != m.end(); ++iter) {
-    headers.push_back(pair<string, string>(iter->first, iter->second));
+  for (const auto& iter: new_env.get_map()) {
```
`iter` -> `kv`.

rzarzynski commented on src/rgw/rgw_rest_swift.cc:
```diff
@@ -571,20 +571,18 @@ static void get_rmattrs_from_headers(const req_state * const s,
   const size_t del_prefix_len = strlen(del_prefix);
 
-  for (iter = m.begin(); iter != m.end(); ++iter) {
+  for (const auto& iter : s->info.env->get_map()) {
```
`iter` -> `kv`.

rzarzynski commented:
@theanalyst: thanks for the update, Abhishek. I think this is the last iteration of review. The changes look good.

theanalyst commented:
@rzarzynski thanks for the patient reviews, updated, also changed a for loop in rgw_rest_client.cc L453 to a range based for loop while i was at it, let me know if this is okay, or I can always fall back to the previous version

rzarzynski commented on src/rgw/rgw_rest_client.cc:
```diff
@@ -264,15 +261,13 @@ int RGWRESTSimpleRequest::forward_request(RGWAccessKey& key, req_info& info, siz
-  for (iter = meta_map.begin(); iter != meta_map.end(); ++iter) {
-    headers.push_back(pair<string, string>(iter->first, iter->second));
+  for (const auto& kv: meta_map) {
+    headers.emplace_back(pair<string, string>(kv.first, kv.second));
```
Maybe:
```cpp
  for (const auto& kv: meta_map) {
    headers.emplace_back(kv);
  }
```

theanalyst replied:
sure, silly me

rzarzynski commented on src/rgw/rgw_rest_client.cc:
```diff
@@ -454,18 +449,15 @@ int RGWRESTStreamWriteRequest::put_obj_init(RGWAccessKey& key, rgw_obj& obj, uin
-  for (bliter = attrs.begin(); bliter != attrs.end(); ++bliter) {
-    bufferlist& bl = bliter->second;
-    const string& name = bliter->first;
+  for (auto attr: attrs) {
```
`auto&`?

theanalyst replied:
done

rzarzynski commented:
jenkins retest this please (`Agent went offline during the build`)

rzarzynski commented:
jenkins retest this please (`cephtool-test-rados.sh`)

rzarzynski commented:
@theanalyst: the Teuthology runs look clean:
* http://pulpito.ceph.com/rzarzynski-2017-06-12_17:02:04-rgw-wip-rzarzynski-testing---basic-mira/,
* http://pulpito.ceph.com/rzarzynski-2017-06-13_13:39:20-rgw-wip-rzarzynski-testing---basic-mira/.

Unfortunately, we've gotten a (trivial) merge conflict in `rgw/rgw_crypt.cc`. Could you please rebase?

theanalyst commented:
@rzarzynski thanks! done

rzarzynski commented:
jenkins retest this please (`test_objectstore_memstore.sh`)
